public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/PyroscopeAppIntegration.tsx

```diff
@@ -0,0 +1,133 @@
+import React, {useEffect, useMemo, useState} from 'react';
+
+import { QueryEditorProps, TimeRange, DateTime } from '@grafana/data';
+import { getBackendSrv } from '@grafana/runtime';
+import {LinkButton} from '@grafana/ui'
+
+import { PyroscopeDataSource } from '../datasource';
+import { PyroscopeDataSourceOptions, Query } from '../types';
+
+const PYROSCOPE_APP_ID = 'grafana-pyroscope-app'
+
+/* Global promise to fetch the pyroscope app settings */
+let pyroscopeAppSettings: Promise<PyroscopeAppSettings> | null= null;
+/* Global promises to fetch pyroscope datasource settings by uid as encountered */
+const pyroscopeDatasourceSettingsByUid: Record<string, Promise<PyroscopeDatasourceSettings>> = {};
+
+export type Props = QueryEditorProps<PyroscopeDataSource, Query, PyroscopeDataSourceOptions>;
+
+/** A subset of the app settings that are relevant for this integration */
+type PyroscopeAppSettings = {
+  enabled: boolean;
+  jsonData: {
+    backendUrl: string;
+    basicAuthUser: string;
+  }
+}
+
+/** A subset of the datasource settings that are relevant for this integration */
+type PyroscopeDatasourceSettings = {
+  url: string;
+  basicAuthUser: string | number; // This might be configured as a number, so we will want to ensure compare for string-equivalence
+}
+
+export function PyroscopeAppIntegration(props: Props) {
+
+  const [appPlugin, setAppPlugin] = useState<PyroscopeAppSettings>()
+  const [datasource, setDatasource] = useState<PyroscopeDatasourceSettings>()
+
+  const {datasource:{uid: datasourceUid}, query, range} = props;
+
+  useEffect(()=>{
+
+    if (pyroscopeAppSettings == null) {
+      pyroscopeAppSettings = getBackendSrv().get<PyroscopeAppSettings>(`/api/plugins/${PYROSCOPE_APP_ID}/settings`)
+    }
+
+    pyroscopeAppSettings.then(setAppPlugin);
+    
+  }, [])
+
+  useEffect(()=>{
+    let datasourceSettings = pyroscopeDatasourceSettingsByUid[datasourceUid];
+
+    if (datasourceSettings == null) {
+      datasourceSettings = getBackendSrv().get<PyroscopeDatasourceSettings>(`/api/datasources/uid/${datasourceUid}`)
+      pyroscopeDatasourceSettingsByUid[datasourceUid] = datasourceSettings;
+    }
+
+    datasourceSettings.then(setDatasource);
+
+  }, [datasourceUid])
+
+  const queryParam = useMemo(()=>{
+
+    return generateQueryParams(query, range);
+  }, [query, range])
+
+  const profilesAppLink = useMemo(()=>isPyroscopeDatasourceCompatibleWithPlugin(datasource, appPlugin) &&
```

Comment:
Why is this memoized? It does not seem to do anything expensive

Reply:
I suppose not. Okay, I'll cut these memos. :+1:

Reply:
Yes, I like the look of the code better without these hooks.